Ensure broadcast when remaining peer becomes unavailable

dirkmc opened the pull request:
Fixes https://github.com/ipfs/go-bitswap/issues/259

dirkmc commented:
The test failure (panic) in CI might be related - we're now telling it to broadcast on disconnect, so that might be having some after-effects. I will dig in.

Stebalien commented on internal/session/sessionwantsender.go:
```diff
@@ -300,7 +307,7 @@ func (spm *sessionWantSender) trackWant(c cid.Cid) {
 }
 
 // processUpdates processes incoming blocks and HAVE / DONT_HAVEs
-func (spm *sessionWantSender) processUpdates(updates []update) {
```
Let's document what this returns.

Stebalien commented on internal/session/sessionwantsender.go:
```diff
@@ -354,21 +361,59 @@ func (spm *sessionWantSender) processUpdates(updates []update) {
+	// (because it may be the last peer who hadn't sent a DONT_HAVE for a CID)
+	if len(newlyUnavailable) > 0 {
+		// Collect all pending wants
+		wants = make([]cid.Cid, len(spm.wants))
```
We just set this above, should we be overwriting it?

dirkmc replied:
The block presence manager has a table of which peers have send a HAVE / DONT_HAVE for each want (CID).

So when we receive a DONT_HAVE, we need to check against the table to see if all peers have now sent us a DONT_HAVE for that want (CID).

In the case where a peer has become unavailable, we need to check against all wants (not just the DONT_HAVEs we just received) because it may be that the peer was the last peer who we were waiting for a response from.

Stebalien replied:
Sure, but what if we have both new `dontHaves` _and_ `newlyUnavailable` peers (specifically, from _different_ peers)?

Stebalien replied:
Oh, nevermind. `wants` is a superset.

Stebalien commented on internal/session/sessionwantsender.go:
```diff
@@ -354,21 +361,59 @@ func (spm *sessionWantSender) processUpdates(updates []update) {
+
+	// If all available peers for a cid sent a DONT_HAVE, signal to the session
+	// that we've exhausted available peers
+	if len(wants) > 0 {
```
All wants from all peers if `newlyUnavailable > 0`?

dirkmc replied:
Let me know if the explanation above makes sense, I think it covers this too.

Stebalien replied:
Ah, I missed the fact that we're _filtering_ this list. This makes sense.

dirkmc commented:
Yeah I agree, there's a lot going on in this class. I've tried to break out bits of it into different classes (eg peerAvailabilityManager) but it still feels too complicated to keep track of everything.

Maybe we can break out the DONT_HAVE tracking into two classes:
- track consecutive DONT_HAVEs (and eject peers that get too many in a row)
- track want exhaustion (every peer sent a DONT_HAVE)

Stebalien commented:
I'm more concerned about threading updates around than having a lot of state. Really, having related state all in one place makes it easier for me to keep track of it.

It's not really an issue with this patch, just something that's creeping up on us.

Stebalien commented on internal/session/sessionwantsender.go:
```diff
@@ -354,21 +362,59 @@ func (spm *sessionWantSender) processUpdates(updates []update) {
+func (spm *sessionWantSender) processExhaustedWants(exhausted []cid.Cid) {
+	newlyExhausted := spm.newlyExhausted(exhausted)
+	if len(newlyExhausted) > 0 {
+		go spm.onPeersExhausted(newlyExhausted)
```
Could we comment on why this needs to be a goroutine?

dirkmc replied:
I will add a comment to the code - but just to clarify here as well:
- `Session` passes `Session.onPeersExhausted()` as a callback into `sessionWantSender`
- `Session.onPeersExhausted()` sends items on the `incoming` channel within `Session`
- The run loop in `Session` pulls items from the `incoming` channel, and the result of processing some of these items is a call to `sessionWantSender`

In summary there is a circular dependency between channels which could lead to a deadlock

Stebalien replied:
Got it. I'd be a _bit_ careful about logical races as this goroutine could be scheduled _much_ later.

I'd also consider moving the goroutine spawn into the callback itself. There, it could _try_ writing to the channel, launching a goroutine only if the write fails.

dirkmc replied:
Huh that's a clever idea, I'll do that

dirkmc commented:
It turns out the problem was with the timing of a test that would verify the session will call FindProviders() after being disconnected from a peer that had the blocks. I changed the logic of the tests to not rely on timing.

While I was doing so I realized that when the session receives "exhausted" wants (wants for whom all peers have sent a DONT_HAVE) it would broadcast a request for *all* pending wants, instead of just those specific exhausted wants. So I also fixed that issue.
